e3db/src/main/java/com/tozny/e3db/Client.java

```diff
@@ -1588,82 +1588,69 @@ public void run() {
    *
    * @param record The record to encrypt.
    * @param eakInfo The key to use for encrypting.
-   * @throws E3DBException An E3DB-specific error occurred.
-   * @return The deccrypted record.
+   * @throws E3DBException Thrown when the signature is not present, the eakInfo does not include a
+   * public signing key, or when verification of the signature fails.
+   * @return The decrypted record.
    */
-  public Record decryptExisting(Record record, EAKInfo eakInfo) throws E3DBException {
-    if (eakInfo.getSignerSigningKey() == null || eakInfo.getSignerSigningKey().isEmpty()) {
-      throw new E3DBException("EAKInfo has no signing key");
-    }
+  public LocalRecord decryptExisting(Record record, EAKInfo eakInfo) throws E3DBException {
+    if(record.signature() == null)
+      throw new E3DBVerificationException(record.meta(), "Record must have a signature in order to decrypt.");
 
-    if (record.signature() == null) {
-      throw new E3DBVerificationException(record.meta());
-    }
+    if (eakInfo.getSignerSigningKey() == null || eakInfo.getSignerSigningKey().isEmpty())
+      throw new IllegalStateException("eakInfo cannot be used to verify the record as it has no public signing key");
 
     byte[] ak = getCachedAccessKey(record, eakInfo);
 
     try {
       Map<String, String> plainRecord = decryptObject(ak, record.data());
 
-      Record decrypted = new LocalRecord(plainRecord, record.meta());
-      SignedDocument<Record> signed = new SD<Record>(decrypted, record.signature());
-
-      if(!verify(signed, eakInfo.getSignerSigningKey())) {
+      if(!verify(new SD<Record>(new LocalRecord(plainRecord, record.meta()),
+          record.signature()), eakInfo.getSignerSigningKey()))
         throw new E3DBVerificationException(record.meta());
-      }
 
-      return new LocalRecord(plainRecord, record.meta());
+      return new LocalRecord(plainRecord, record.meta(), record.signature());
     } catch (UnsupportedEncodingException e) {
       throw new RuntimeException(e);
     }
   }
 
   /**
-   * Encrypt an existing record for local storage.
+   * Sign &amp; encrypt an existing record for local storage.
    *
    * @param record The record to encrypt.
    * @param eakInfo The key to use for encrypting.
-   * @return The encrypted record. Note that server-provided metadata values will not throw if accessed.
+   * @return The encrypted record.
    */
-  public Record encryptExisting(Record record, EAKInfo eakInfo) {
+  public EncryptedRecord encryptExisting(Record record, EAKInfo eakInfo) {
     checkNotNull(record, "record");
     checkNotNull(eakInfo, "eak");
 
-    String signature = record.signature();
-    if (signature == null && Client.this.privateSigningKey != null) {
-      SignedDocument<Record> document = sign(record);
-      signature = document.signature();
-    }
+    if(Client.this.privateSigningKey == null)
+      throw new IllegalStateException("Client must have a signing key to encrypt locally.");
```

Comment:
Makes me think that we should somehow allow a "late" registration of public signing key, but IIRC the mobile SDKs landed with signing keys? Maybe not...